diff --git a/lib/Uppy.js b/lib/Uppy.js
index ca5f906412a8057f2b8bdaec11369152a01837b9..c14fc900c6fa615604300e043b12e3e6a99a0d28 100644
--- a/lib/Uppy.js
+++ b/lib/Uppy.js
@@ -26,6 +26,7 @@ export class Uppy {
     #plugins = Object.create(null);
     #restricter;
     #storeUnsubscribe;
+    #onlineStatusTimeout;
     #emitter = ee();
     #preProcessors = new Set();
     #uploaders = new Set();
@@ -1128,7 +1129,7 @@ export class Uppy {
         if (typeof window !== 'undefined' && window.addEventListener) {
             window.addEventListener('online', this.#updateOnlineStatus);
             window.addEventListener('offline', this.#updateOnlineStatus);
-            setTimeout(this.#updateOnlineStatus, 3000);
+            this.#onlineStatusTimeout = setTimeout(this.#updateOnlineStatus, 3000);
         }
     }
     updateOnlineStatus() {
@@ -1246,6 +1247,8 @@ export class Uppy {
      * Uninstall all plugins and close down this Uppy instance.
      */
     destroy() {
+        clearTimeout(this.#onlineStatusTimeout);
+        this.#onlineStatusTimeout = undefined;
         this.log(`Closing Uppy instance ${this.opts.id}: removing all files and uninstalling plugins`);
         this.cancelAll();
         this.#storeUnsubscribe();
diff --git a/src/Uppy.ts b/src/Uppy.ts
index 0aa2ffefb389effb85457770575d90753013ee63..6620de2c4cfb07639ab6bc0dcd55dd00cbbd1ace 100644
--- a/src/Uppy.ts
+++ b/src/Uppy.ts
@@ -417,6 +417,8 @@ export class Uppy<
 
   #storeUnsubscribe
 
+  #onlineStatusTimeout: ReturnType<typeof setTimeout> | undefined
+
   #emitter = ee()
 
   #preProcessors: Set<Processor> = new Set()
@@ -1885,7 +1887,7 @@ export class Uppy<
     if (typeof window !== 'undefined' && window.addEventListener) {
       window.addEventListener('online', this.#updateOnlineStatus)
       window.addEventListener('offline', this.#updateOnlineStatus)
-      setTimeout(this.#updateOnlineStatus, 3000)
+      this.#onlineStatusTimeout = setTimeout(this.#updateOnlineStatus, 3000)
     }
   }
 
@@ -2041,6 +2043,9 @@ export class Uppy<
    * Uninstall all plugins and close down this Uppy instance.
    */
   destroy(): void {
+    clearTimeout(this.#onlineStatusTimeout)
+    this.#onlineStatusTimeout = undefined
+
     this.log(
       `Closing Uppy instance ${this.opts.id}: removing all files and uninstalling plugins`,
     )
